dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataverse/themewidget/ThemeWidgetFragment.java

```diff
@@ -279,27 +278,30 @@ public boolean isThemeEnabled() {
 
 
     public String save() {
-        // If this Dv isn't the root, delete the uploaded file and remove theme
-        // before saving.
-        if (!editDv.isThemeRoot()) {
-            uploadedFile = null;
-            editDv.setDataverseTheme(null);
+        Try<Dataverse> saveOperation;
+
+        if (editDv.isThemeRoot()) {
+            saveOperation = Try.of(() -> themeWidgetService.saveOrUpdateUploadedTheme(editDv, uploadedFile));
+        } else {
+            saveOperation = Try.of(() -> themeWidgetService.inheritThemeFromRoot(editDv))
+                    .onSuccess(editedDataverse -> editDv = editedDataverse);
         }
-        Command<Dataverse> cmd = new UpdateDataverseThemeCommand(editDv, this.uploadedFile, dvRequestService.getDataverseRequest());
-        try {
-            commandEngine.submit(cmd);
-        } catch (Exception ex) {
-            logger.log(Level.SEVERE, "error updating dataverse theme", ex);
-            FacesContext.getCurrentInstance().addMessage(null, new FacesMessage(FacesMessage.SEVERITY_FATAL, BundleUtil.getStringFromBundle("dataverse.save.failed"), BundleUtil.getStringFromBundle("dataverse.theme.failure")));
+
+        saveOperation.andThen(dataverse -> this.cleanupTempDirectory());
+
+        if (saveOperation.isSuccess()) {
+            JsfHelper.addFlashSuccessMessage(BundleUtil.getStringFromBundle("dataverse.theme.success"));
+            return "dataverse.xhtml?faces-redirect=true&alias=" + editDv.getAlias();
+        } else {
+            logger.log(Level.SEVERE, "error updating dataverse theme", saveOperation.getCause());
+            FacesContext.getCurrentInstance().addMessage(null,
```

Comment:
We should use JsfHelper, it could help in the future if we would have some problems again with jsf error messages, and we would need to "hack" something inside "add jsf message" method.